Issue #919 - Stop debug vm on end of debug session.

rorygraves opened the pull request:
Debug sessions launched by Ensime were not terminated when Ensime finishes debugging.
This commit makes the debugged vm stop when we terminate the debug vm session.
This may be a little agressive for attached sessions.

aemoncannon commented:
Cool

On Tue, May 5, 2015, 1:20 AM Rory <notifications@github.com> wrote:

> rorygraves wants to merge 1 commit into ensime:master from
> rorygraves:stop_vm_should_stop:
>
> Debug sessions launched by Ensime were not terminated when Ensime finishes
> debugging.
> This commit makes the debugged vm stop when we terminate the debug vm
> session.
> This may be a little agressive for attached sessions.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/ensime/ensime-server/pull/957
> Commit Summary
>
>    - Issue #919 - Stop debug vm on end of debug session.
>
> File Changes
>
>    - *M* server/src/main/scala/org/ensime/core/DebugManager.scala
>    <https://github.com/ensime/ensime-server/pull/957/files#diff-0> (5)
>
> Patch Links:
>
>    - https://github.com/ensime/ensime-server/pull/957.patch
>    - https://github.com/ensime/ensime-server/pull/957.diff
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ensime/ensime-server/pull/957>.
>

edani commented:
@rorygraves  I think for attached sessions, exit() shouldn't be called, just dispose().

rorygraves commented:
@edani Yeah fair comment - will look at it when i get the chance.

rorygraves commented:
lol, went to fix this over the weekend and found several other bugs - so bigger refactor in progress - closing for now.

rorygraves commented on core/src/main/scala/org/ensime/core/DebugManager.scala:
```diff
@@ -237,67 +239,75 @@ class DebugManager(
     case DebugStopReq =>
-      handleRPCWithVM() { vm =>
+      sender ! handleRPCWithVM() { vm =>
+        if (vm.mode.shouldExit) {
```
@edani This fix means it should only trigger the exit on a VM started by Ensime - as per original discussion.

rorygraves commented:
@fommil @edani @aemoncannon 
So this is a rework of the original fix - got a bit carried away, but DebugManager definately has less errors in it right now.  Fundamental change is that the ```sender !``` only occurs in the Receive block and everything else returns an RpcResponse and the type system helps to ensure this is the case.  I found a few cases where a non-response value or, in one case, no value would be returned.
So in summary, fixed the stop thing, cleaned up a bunch of other stuff...

edani commented:
@rorygraves   This looks good to me. I'll see if I can reproduce the appveyor failure locally.

fommil commented:
Should we stop all VM that we started in the post Stop ?

rorygraves commented:
Excellent point. Yes.

rorygraves commented:
Thanks @edani
